main/src/ca/uwaterloo/flix/language/phase/Monomorpher.scala

```diff
@@ -520,6 +520,15 @@ object Monomorpher {
       val es = exps.map(visitExp(_, env0, subst))
       MonoAst.Expr.Apply(e, es, subst(tpe), subst(eff), loc)
 
+    case LoweredAst.Expr.ApplyDef(Ast.DefSymUse(sym, loc1), exps, tpe, eff, loc2) =>
+      val spec = root.defs(sym).spec
+      val declaredTypes = spec.fparams.map(_.tpe)
+      // Is this correct or should this be curried? Maybe done in ConstraintGen?
+      val tpe1 = Type.mkApply(Type.Cst(TypeConstructor.Arrow(declaredTypes.length), loc1), spec.eff :: declaredTypes, loc1)
```

Comment:
Looks like you're missing a return type here. There's something like Type.mkArrowWithEffect, which you may find useful